src/scripts/generateEmissionsTable.ts

```diff
@@ -0,0 +1,367 @@
+    gasUsedArray.push(gas[i]);
+    timestampArray.push(gas[i].UnixTimeStamp);
+  }
+  return [gasUsedArray, timestampArray];
+};
+
+export const findClosest = (goal: number, array: any[]) => {
+  let closest = array[0];
+  let diff = Math.abs(goal - closest);
+  for (let i = 0; i < array.length; i++) {
+    let newdiff = Math.abs(goal - array[i]);
+    if (newdiff < diff) {
+      diff = newdiff;
+      closest = array[i];
+    }
+  }
+  return closest;
+};
+
+export const fetchIndexesFromBlockResolution = async (
+  blockResolution: number,
+  currentBlock: number,
+  gas: networkData[]
+) => {
+  console.log("Generating index array at block resolution");
+  //Need to arrayify data to identify which is closest
+  const [gasUsedArray, timestampArray] = arrayifyCSVData(gas);
+
+  let nextBlock = blockResolution;
+
+  let indexArray = [];
+
+  for (let i = 0; i < currentBlock; i++) {
+    if (i % blockResolution === 0) {
+      nextBlock += blockResolution;
+
+      //Find nearest UNIX timestamp within csv data to the input block
+      const response = await web3.eth.getBlock(i);
+
+      let goal = 0;
+
+      if (typeof response.timestamp === "number") {
+        goal = response.timestamp;
+      }
+
+      const closest = findClosest(goal, timestampArray);
+
+      /*
+            let closest = timestampArray.reduce(function(prev:number, curr:number) {
+                return (Math.abs(curr - goal) < Math.abs(prev - goal) ? curr : prev);
+            });
+            */
+
+      //Find index of closest csv data point
+      let index = gasUsedArray.findIndex(
+        (x: any) => x.UnixTimeStamp === closest
+      );
+
+      indexArray.push(
+        new (blockData as any)(index, gasUsed[index].UnixTimeStamp, i)
+      );
+    }
+  }
+
+  //Append final value from JSON
+  const finalIndex = gasUsed.length - 1;
+  indexArray.push(
+    new (blockData as any)(
+      finalIndex,
+      gasUsed[finalIndex].UnixTimeStamp,
+      currentBlock
+    )
+  );
+  console.log(indexArray);
+  return indexArray;
+};
+
+export const fetchIndexesFromDayResolution = async (
+  dayResolution: number,
+  gas: networkData[]
+) => {
+  let indexArray = [];
+  console.log("Generating index array at day resolution");
+  //Loop through gas used data
+  for (let i = 0; i < gas.length; i++) {
+    //If we are at the start of the day range, push that index data to array
+    if (i % dayResolution === 0) {
+      // Catch any timestamps before block 1
+      let UNIXTimestamp = gas[i].UnixTimeStamp;
+      if (UNIXTimestamp < 1438270000) {
+        UNIXTimestamp = 1438270000;
+      }
+
+      // Find block number from timestamp
+      // Construct etherscan URL
+      const etherscanURL =
+        "https://api.etherscan.io/api?module=block&action=getblocknobytime&timestamp=" +
+        UNIXTimestamp +
+        "&closest=before&apikey=" +
+        ETHERSCAN_API_KEY;
+
+      // Fetch etherscan data
+      async function fetchEtherscanData(url: string) {
```

Comment:
To mock this in a test, what we need to do is put it in a separate .ts file as a utility function, and import it into this main script file. Then you can mock the whole import at test runtime, here is an example:
https://github.com/Offsetra/ethereum-emissions-calculator/blob/a0b80f277b6bea342e9f5b41e43306106f555dc6/src/index.test.ts#L9

But if you do want to write tests for the util file, you will still need to mock `fetch`. Here is an example:
https://github.com/Offsetra/ethereum-emissions-calculator/blob/bd7d995670b1f256aa264107b1fa58875ccbcd2c/src/utils/fetchJSON.test.ts#L22

because its global, be sure to clear the mocks as they do above